workload/schemachange: fix CREATE FUNCTION flakes

annrpom opened the pull request:
### workload/schemachange: limit amount of params generated by create fn
Previously, our schemachange workload flaked due to some `CREATE
FUNCTION`s 
randomly generated with >= 100 params; however, this is not
allowed and not a particularly useful edge case to test. This patch
prevents these types of functions from being generated by adding a
restriction to the max length of fn params.

Informs: #116649

Epic: none
Release note: None

---

### workload/schemachange: handle cases of duplicate functions
This patch ensures that our schemachange workload is able
to handle cases where we could generate duplicate functions
(same name, same param types). We will now detect whether
or not a duplicate function has been generated and add the
proper error to our expected errors list.

Fixes: #117421

Epic: none
Release note: None

annrpom committed:
workload/schemachange: limit amount of params generated by create fn

Previously, our schemachange workload flaked due to some `CREATE
FUNCTION`s randomly generated with >= 100 params; however, this is not
allowed and not a particularly useful edge case to test. This patch
prevents these types of functions from being generated by adding a
restriction to the max length of fn params.

Informs: #116649

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/117432)
<!-- Reviewable:end -->

rafiss commented on pkg/workload/schemachange/operation_generator.go:
```diff
@@ -3921,7 +3922,7 @@ func (og *operationGenerator) createFunction(ctx context.Context, tx pgx.Tx) (*o
 		},
 		"ParamRefs": func() (string, error) {
-			refs, err := PickAtLeast(og.params.rng, 1, possibleParamReferences)
+			refs, err := PickBetween(og.params.rng, 1, optbuilder.MaxFuncParams-1, possibleParamReferences)
```
i think we could make this avoid the `optbuilder` dependency, and just hardcode something like `20` here. it would be better for readability too.

annrpom committed:
workload/schemachange: handle cases of duplicate functions

This patch ensures that our schemachange workload is able
to handle cases where we could generate duplicate functions
(same name, same param types). We will now detect whether
or not a duplicate function has been generated and add the
proper error to our expected errors list.

Fixes: #117421

Epic: none
Release note: None

annrpom commented on pkg/workload/schemachange/error_screening.go:
```diff
@@ -98,6 +98,18 @@ func (og *operationGenerator) schemaExists(
+func (og *operationGenerator) fnExists(
+	ctx context.Context, tx pgx.Tx, fnName string,
+) (bool, error) {
+	// TODO(before merge): we want to use specfic_name
```
@rafiss I am still looking at this, but just in case you know off of the top of your head: is there a way we can check if the newly created fn is a duplicate by name and param types of any other existing fn?

using specific_name sounds nice, but i am not too sure on how we get from the template SQL string to its specific_name

rafiss replied:
this could be a good starting point:
```
select proname, proargnames from pg_proc where proname = 'f';
```

(technically that just matches on the names. to check types also, it would be in `proargtypes`)

annrpom replied:
gotcha - it just looks like the set of param types matters here; so, i just focused on that! thank you

annrpom commented on pkg/workload/schemachange/operation_generator.go:
```diff
@@ -3941,13 +3928,47 @@ func (og *operationGenerator) createFunction(ctx context.Context, tx pgx.Tx) (*o
+	// Upon a successful completion, let's check to make sure that our
+	// function doesn't exist already (overloads are fine).
+	if expectedCode == pgcode.SuccessfulCompletion {
+		params := util.Map(stmt.Params, func(t tree.RoutineParam) string {
```
👋 i can also add `fold` (i think also known as `reduce`) to this utils pkg as well instead of mapping, but figured i would suggest it first before doing (i promise it would be quick)

renatolabs replied:
My 2c: I think the `map` implementation is fine. With `fold`, the closure would have to manage the formatting inline, and it's nice to rely on `strings.Join` for that.

renatolabs commented on pkg/workload/schemachange/operation_generator.go:
```diff
@@ -3921,7 +3921,7 @@ func (og *operationGenerator) createFunction(ctx context.Context, tx pgx.Tx) (*o
 		},
 		"ParamRefs": func() (string, error) {
-			refs, err := PickAtLeast(og.params.rng, 1, possibleParamReferences)
+			refs, err := PickBetween(og.params.rng, 1, 20, possibleParamReferences)
```
> this is not allowed and not a particularly useful edge case to test

While I agree that testing > 100 params is not particularly interesting, what I think _is_ interesting is testing functions with exactly the maximum number of parameters, or close to that amount.

In terms of state space exploration, I think picking `{1, 20}` is not optimal. We're 20x-ing our state space but arguably going from a function with 2 params to a function with 3 params, as an example, is likely to not uncover different paths or behaviour in the DB.

IMO, a more efficient way to go would be to `PickOne` out of a few "classes" of functions:

* lowest possible number of params (1? 0?)
* few parameters (say, 3)
* "medium" amount of parameters (say 10)
* large amount of parameters (say 60)
* max allowed number of parameters (100?)

We now only have 5 paths that better cover our ability to handle different numbers of parameters in crdb.

Related question: is there a reason we don't currently test functions with 0 params?

annrpom replied:
hihi thanks for taking a look 🙇  - I think you make a good point, but I am hesitant to say if any of those cases would really vary in behavior (in terms of code coverage). However, if anything, we could bump up 20 to be 99 and rng could cover those "classes" to a reasonable extent

Rafi might also chime in if he would like; it's very possible I'm overlooking something here

Also, we do test functions with 0 params: the "ParamRefs" key in our `placeholderMap` funcmap randomizes between using params vs not by `useParamRefs` https://github.com/cockroachdb/cockroach/blob/1898adc059e9b3026fd5a7ed88dc41cbb006abfc/pkg/workload/schemachange/operation_generator.go#L3878

used in this way

https://github.com/cockroachdb/cockroach/blob/1898adc059e9b3026fd5a7ed88dc41cbb006abfc/pkg/workload/schemachange/operation_generator.go#L3925-L3928

renatolabs replied:
> I am hesitant to say if any of those cases would really vary in behavior (in terms of code coverage)

I can understand that we might not be necessarily increasing source code coverage by much, but it does increase coverage of state space in the DB and that's what really matters. Testing functions with higher number of params would also make sure that they are handled properly outside of schema land too -- things like backup/restores, upgrade migrations, etc.

Using `PickBetween(min, max)` could be something we do as well, although I still prefer the more limited choice set, as that would test edge cases more often.

Thanks for explaining how the "no params" part of the workload works!

renatolabs commented on pkg/workload/schemachange/operation_generator.go:
```diff
@@ -3941,13 +3928,47 @@ func (og *operationGenerator) createFunction(ctx context.Context, tx pgx.Tx) (*o
+
 	return newOpStmt(stmt, codesWithConditions{
 		{expectedCode, true},
+		{code: pgcode.DuplicateFunction, condition: fnDuplicate},
```
Nit: for consistency, I'd make this line formatted the same way as the line above (i.e., either change this line to be implicit or the line above to be explicit).

renatolabs reviewed (approved):
Nice fix! All my comments are optional.

rafiss reviewed (approved):
nice work!

> Reviewed 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117432#-:-NoTrXF36qOgNiYL-mA7:b-j4xmku)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom and @DarrylWong)



<!-- Sent from Reviewable.io -->

annrpom commented:
TFTRs! ('-')7

bors r=renatolabs, rafiss
